docs/04-reference/winglang-spec.md

```diff
@@ -869,30 +869,223 @@ let f = (arg1: num, var arg2: num) => {
 
 ### 1.7 Optionality
 
-Symbol `?` can mark a type as optional.  
-Optionality means the value behind type can be either present or nil.
+Nullity (`null`, `undefined`, `nil`) is a primary source of bugs in software.
+Being able to guarantee that a value will never be null makes it easier to write safe
+code without constantly having to take nullity into account.
 
-Rules of optionality apply to the entire new container type of `type?` and not
-the value behind it (`type`).  
+We are going to try and design Wing without allowing users to explicitly assign or use the `nil`
+value. Instead, the language offers syntax to support the various use cases that arise when dealing
+with *optionality*.
 
-Using the `??` operator allows one to safely access the value behind an optional
-variable by always providing a default, in case the variable is nil. This forces
-a value to be present for the l-value (left hand side of the assignment operator
-and guarantees what's returned is type-stripped from its `?` keyword).
+Here's a quick summary of how optionality works in Wing:
 
-> ```TS
-> let x? = 44;
-> let y = x ?? 55;
-> ```
+* `x: T?` marks `x` as "optional of T". This means that `x` can store a value of type `T` but it can also be in a state where there is no value stored there.
```

Comment:
I am trying to understand what is the idiom for lazy instantiation in the current version
```ts (wing) 
resource {
  _hashcode: str? 
  _compute_hashcode() : str {
  } 

  hashcode(): str { 
    if let var h = this._hashcode {
      return h;
    } else {
      h = this.calculate_hashcode();
      this._hashcode = h;
      return h;
    } 
  }
} 
```
The above should work, but it is very verbose

Reply:
Alternatively with `let/else` (note that the field needs to be marked `var`):

```js
resource Foo {
  var _hashcode: str?;

  hashcode() {
    let hc = this._hashcode else {
      this._hashcode = this._compute_hashcode();
      return this.hashcode();
    };

    return hc;
  }
}
```

Eventually, this is why we need `??=`:

```js
resource Foo {
  var _hashcode: str?;
  
  hashcode() {
    return this._hashcode ??= this._compute();
  }
}
```